File: swift/src/Calculator.swift
import Foundation

// Project: Calculator
// Practice: Functions, control flow, optional return types

func calculate(a: Double, b: Double, operation: String) -> Double? {
    switch operation {
    case "+":
        return a + b
    case "-":
        return a - b
    case "*":
        return a * b
    case "/":
        return b != 0 ? a / b : nil
    case "%":
        return b != 0 ? a.truncatingRemainder(dividingBy: b) : nil
    default:
        return nil
    }
}

func printMenu() {
    print("""
    === Swift Calculator ===
    Select operation:
    + : Addition
    - : Subtraction
    * : Multiplication
    / : Division
    % : Modulus
    """)
}

print("=== Swift Calculator Demo ===\n")

// Demo calculations
let num1 = 15.0
let num2 = 3.0

let operations = ["+", "-", "*", "/", "%"]

for op in operations {
    if let result = calculate(a: num1, b: num2, operation: op) {
        print("\(num1) \(op) \(num2) = \(result)")
    }
}

// Division by zero handling
print("\nDivision by zero test:")
if let result = calculate(a: 10, b: 0, operation: "/") {
    print("Result: \(result)")
} else {
    print("Error: Cannot divide by zero")
}

// Interactive mode (uncommented for manual testing)
/*
print("\nEnter interactive mode? (Type: yes/no)")
if let response = readLine()?.lowercased(), response == "yes" {
    printMenu()
    
    print("Enter first number: ", terminator: "")
    guard let aStr = readLine(), let a = Double(aStr) else {
        print("Invalid input")
        exit(1)
    }
    
    print("Enter operation: ", terminator: "")
    guard let op = readLine() else {
        print("Invalid input")
        exit(1)
    }
    
    print("Enter second number: ", terminator: "")
    guard let bStr = readLine(), let b = Double(bStr) else {
        print("Invalid input")
        exit(1)
    }
    
    if let result = calculate(a: a, b: b, operation: op) {
        print("\nResult: \(result)")
    } else {
        print("\nInvalid operation or error in calculation")
    }
}
*/

File: kotlin/src/Calculator.kt
// Project: Calculator
// Practice: Functions, when expressions, null safety

fun calculate(a: Double, b: Double, operation: String): Double? {
    return when (operation) {
        "+" -> a + b
        "-" -> a - b
        "*" -> a * b
        "/" -> if (b != 0.0) a / b else null
        "%" -> if (b != 0.0) a % b else null
        else -> null
    }
}

fun printMenu() {
    println("""
        === Kotlin Calculator ===
        Select operation:
        + : Addition
        - : Subtraction
        * : Multiplication
        / : Division
        % : Modulus
    """.trimIndent())
}

fun main() {
    println("=== Kotlin Calculator Demo ===\n")
    
    // Demo calculations
    val num1 = 15.0
    val num2 = 3.0
    
    val operations = listOf("+", "-", "*", "/", "%")
    
    for (op in operations) {
        val result = calculate(num1, num2, op)
        if (result != null) {
            println("$num1 $op $num2 = $result")
        }
    }
    
    // Division by zero handling
    println("\nDivision by zero test:")
    val result = calculate(10.0, 0.0, "/")
    if (result != null) {
        println("Result: $result")
    } else {
        println("Error: Cannot divide by zero")
    }
    
    // Functional approach
    println("\nBatch calculations:")
    val pairs = listOf(
        Triple(20.0, 4.0, "+"),
        Triple(100.0, 10.0, "-"),
        Triple(7.0, 6.0, "*")
    )
    
    pairs.forEach { (a, b, op) ->
        val res = calculate(a, b, op)
        println("$a $op $b = ${res ?: "Error"}")
    }
}
